examples/tonic/src/server.rs

```diff
@@ -0,0 +1,102 @@
+use tracing_subscriber::{Layer, Registry};
+
+pub mod hello_world {
+    tonic::include_proto!("helloworld"); // The string specified here must match the proto package name
+}
+
+#[derive(Debug, Default)]
+pub struct MyGreeter {}
+
+#[tonic::async_trait]
+impl Greeter for MyGreeter {
+    async fn say_hello(
+        &self,
+        request: Request<HelloRequest>, // Accept request of type HelloRequest
+    ) -> Result<Response<HelloReply>, Status> {
+        // Return an instance of type HelloReply
+        println!("Got a request: {:?}", request);
+
+        let reply = hello_world::HelloReply {
+            message: format!("Hello {}!", request.into_inner().name), // We must use .into_inner() as the fields of gRPC requests and responses are private
+        };
+
+        Ok(Response::new(reply)) // Send back our formatted greeting
+    }
+}
+
+fn tracing_init() -> Result<(), Box<dyn std::error::Error>> {
+    let builder = opentelemetry_jaeger::Exporter::builder()
+        .with_agent_endpoint("127.0.0.1:6831".parse().unwrap());
+
+    let exporter = builder
+        .with_process(opentelemetry_jaeger::Process {
+            service_name: "server".to_string(),
+            tags: vec![KeyValue::new("version", "0.1.0")],
+        })
+        .init()?;
+
+    let provider = sdk::Provider::builder()
+        .with_simple_exporter(exporter)
+        .with_config(sdk::Config {
+            default_sampler: Box::new(Sampler::Always),
+            ..Default::default()
+        })
+        .build();
+
+    let tracer = provider.get_tracer("my-tracer");
+    let opentelemetry = OpenTelemetryLayer::with_tracer(tracer);
+
+    let subscriber = Registry::default()
+         // add the OpenTelemetry subscriber layer
+        .with(opentelemetry)
+         // add a logging layer
+        .with(tracing_subscriber::fmt::Layer::default())
+         // add RUST_LOG-based filtering
+        .with(tracing_subscriber::EnvFilter::from_default_env());
+    tracing::subscriber::set_global_default(subscriber)?;
+    global::set_provider(provider);
+
+    Ok(())
+}
+
+struct HttpHeaderMapCarrier<'a>(&'a http::HeaderMap);
+impl<'a> api::Carrier for HttpHeaderMapCarrier<'a> {
+    fn get(&self, key: &'static str) -> Option<&str> {
+        self.0
+            .get(key.to_lowercase().as_str())
+            .and_then(|value| value.to_str().ok())
+    }
+
+    fn set(&mut self, _key: &'static str, _value: String) {
+        unimplemented!()
+    }
+}
+
+#[tokio::main]
+async fn main() -> Result<(), Box<dyn std::error::Error>> {
+    tracing_init()?;
+    let addr = "[::1]:50051".parse()?;
+    let greeter = MyGreeter::default();
+    let propagator = api::B3Propagator::new(true);
+
+    Server::builder()
+        .trace_fn(move |header| {
+            let parent = propagator.extract(&HttpHeaderMapCarrier(header));
+            let tracing_span = tracing::info_span!("Received request");
+            tracing_span.set_opentelemetry_parent(parent);
```

Comment:
I'm not sure if this is an issue with the contract of HttpTextFormat, or with how set_opentelemetry_parent handles an empty SpanContext, but setting this when there aren't any headers leads to each span ending up in its own trace ID somehow.

Reply:
Without set_opentelemetry_parent:
![image](https://user-images.githubusercontent.com/5514447/78195352-47cd9c00-7434-11ea-8530-bcd25cfa9fc9.png)

set_opentelemetry_parent with an empty SpanContext
![image](https://user-images.githubusercontent.com/5514447/78195320-35ebf900-7434-11ea-8010-028be2f050f4.png)

Reply:
Yeah there are a few outstanding PRs still in the core `opentelemetry` crate that should fix this. I could do a better job of cross-linking, primarily waiting on https://github.com/open-telemetry/opentelemetry-rust/pull/84

Reply:
This should now be fixed after moving to opentelemetry 0.4.0